### PR TITLE
Update to OpenSSL 3 by default

### PR DESCRIPTION
**Pull request by cjones051073**
- update openssl shim port and PG to use openssl 3 by default.
- rev bump all ports depending on openssl (using attached script).

Closes: [63801 - python ports: update to use openssl 3, allowing dependent binaries to be redistributed](https://trac.macports.org/ticket/63801)

[macports-bump-rev.sh.txt](https://github.com/macports/macports-ports/files/7478378/macports-bump-rev.sh.txt)

### CODE REVIEW

**Review by mascguy — APPROVED** [review]

**Comment by jmroot** [comment]
Shouldn't only the ports using the default openssl version be rev bumped?

**Comment by cjones051073** [comment]
> Shouldn't only the ports using the default openssl version be rev bumped?

my script rev-bumps ports that depend on the openssl shim port. nothing more nothing less...

**Comment by cjones051073** [comment]
> > Shouldn't only the ports using the default openssl version be rev bumped?
> 
> my script rev-bumps ports that depend on the openssl shim port. nothing more nothing less...

to be clear, I have no planes to do any more or less than my script does. if that rev bumps ports that do not need it (unlikely) or misses ports, so be it..

**Comment by mascguy** [comment]
In terms of the logistics of ultimately merging this, do we expect the buildbots to handle this without dying?

**Comment by cjones051073** [comment]
> In terms of the logistics of ultimately merging this, do we expect the buildbots to handle this without dying?

no. they will always certainly timeout...

> At the very least, the watchers are going to be tied up for a LONG time, processing the full dependency list...

ultimately I believe it comes down to a judgement call. do we 'merge and deal with the aftermath as is' or .... personally I am not sure there is much of an or...

**Comment by cjones051073** [comment]
> > In terms of the logistics of ultimately merging this, do we expect the buildbots to handle this without dying?
> 
> no. they will always certainly timeout...

Just be 100%B clear I assume you are referring to the CI test builds here. these will probably timeout. The macports buildbots will handle then update though just fine...

**Comment by mascguy** [comment]
> > > In terms of the logistics of ultimately merging this, do we expect the buildbots to handle this without dying?
> > 
> > 
> > no. they will always certainly timeout...
> 
> Just be 100% clear, I assume you are referring to the CI test builds here ? these will probably timeout. The macports buildbots will handle then update though just fine...

Nope, I'm talking about the buildbots. If it's one giant commit, the watchers will be tied up for an eternity, whilst leaving the builders idle.

**Comment by mascguy** [comment]
My recommendation - barring more definitive guidance from Josh and Ryan - would be to limit the number of rev-bumps to no more than 50 per commit.

**Comment by cjones051073** [comment]
> My recommendation - barring more definitive guidance from Josh and Ryan - would be to limit the number of rev-bumps to no more than 50 per commit.
> 
> And while we're not dealing with thousands of rev-bumps, I'm concerned we might have an issue like this from a few months ago:
> 
> https://lists.macports.org/pipermail/macports-dev/2021-June/043605.html

I am not going to do that. I don’t thing this PR is that large that merging it as is is that much of an issue.

**Comment by jmroot** [comment]
> my script rev-bumps ports that depend on the openssl shim port. nothing more nothing less...

That doesn't appear to be the case. For example python26 is bumped and it only depends on openssl10.

**Comment by cjones051073** [comment]
> > my script rev-bumps ports that depend on the openssl shim port. nothing more nothing less...
> 
> That doesn't appear to be the case. For example python26 is bumped and it only depends on openssl10.

My script is probably not perfect. Bar doing it by hand, which I am not going to do, its better than nothing.

**Comment by cjones051073** [comment]
> > my script rev-bumps ports that depend on the openssl shim port. nothing more nothing less...
> 
> That doesn't appear to be the case. For example python26 is bumped and it only depends on openssl10.



> > > my script rev-bumps ports that depend on the openssl shim port. nothing more nothing less...
> > 
> > 
> > That doesn't appear to be the case. For example python26 is bumped and it only depends on openssl10.
> 
> My script is probably not perfect. Bar doing it by hand, which I am not going to do, its better than nothing.



> > > my script rev-bumps ports that depend on the openssl shim port. nothing more nothing less...
> > 
> > 
> > That doesn't appear to be the case. For example python26 is bumped and it only depends on openssl10.
> 
> My script is probably not perfect. Bar doing it by hand, which I am not going to do, its better than nothing.

```
Oberon ~/Downloads > port echo depends:openssl | grep python
python26                        
python27                        
python27-bootstrap              
python32                        
python33                        
python34                        
python35                        
python36                        
python37                        
python38                        
python39                        
python310                       
python311-devel 
```

if python26 should not appear in the list above, thats a bug in base.

**Comment by jmroot** [comment]
It's a bug in your regex. Try `depends:':openssl(\s|$)'`.

**Comment by mascguy** [comment]
Hey Chris, bear in mind that we're not trying to be critical of what you've done here. And we appreciate you taking the initiative, which is often the most important step!

Others are happy to assist, to help refine all of this. Whether that means breaking up the commits into smaller chunks, trimming the list of rev-bumps, or whatever. As long as you don't block commits to this PR by your fellow members, we'll help too!

**Comment by cjones051073** [comment]
> It's a bug in your regex. Try `depends:':openssl(\s|$)'`.

thanks. TBH I still view this as a conceptual bug in base that I have to do this to avoid unintentional matches...

**Comment by cjones051073** [comment]
> Hey Chris, bear in mind that we're not trying to be critical of what you've done here. And we appreciate you taking the initiative, which is often the most important step!
> 
> Others are happy to assist, to help refine all of this. Whether that means breaking up the commits into smaller chunks, trimming the list of rev-bumps, or whatever. As long as you don't block commits to this PR by your fellow members, we'll help too!
> 
> p.s. We love ya Man!

thats fine, but just to be 100% clear I am not going to waste time myself breaking this up into smaller MRs. I don't see this as being necessary. I generated the changes here by a) updating the openssl Pam to default to 3 and b) running my attached script to auto-rev bump dependent ports. thats all I am prepared to do, but you want more fine grained than that please take the script and knock yourself out ;)...

**Comment by kencu** [comment]
Well well well. Deja vu.

**Comment by Tatsh** [comment]
I am approving of this OpenSSL update for OpenRCT2 😄

**Comment by RJVB** [comment]
On Thursday November 04 2021 15:36:15 Chris Jones wrote:

>thats fine, but just to be 100% clear I am not going to waste time myself breaking this up into smaller MRs.

I suppose one could get the rev-bump approval here, but drop the actual change like it's already been done elsewhere AFAIK. If shouldn't take a lot of time to adapt your script so it applies only to approved ports, and applies the changes to your working copy of the actual ports tree. It could even do the pushing at some kind of interval (but I doubt the build bots will be less stressed by a series of commits than a single big commit).

What I'm missing is a little bit of background. I wasn't even aware of OpenSSL3 (what ever happened to OpenSSL2?), only vaguely about the OpenSSL PG. I think the least would be to touch a few words on why the rev-bump is necessary. Is it because of an ABI incompatibility or some potentially less stringent reason?

**Comment by lhaeger** [comment]
https://wiki.openssl.org/index.php/OpenSSL_3.0#Main_Changes_in_OpenSSL_3.0_from_OpenSSL_1.1.1

> Am 05.11.2021 um 09:55 schrieb René Bertin ***@***.***>:
> 
> What I'm missing is a little bit of background.

**Comment by ryandesign** [comment]
> > In terms of the logistics of ultimately merging this, do we expect the buildbots to handle this without dying?
> 
> no. they will almost certainly timeout...

Buildbot will not time out. The only situation in which our buildbot will time out a portbuilder build is if it goes for 60 minutes without producing any output, and I'm not aware of any ports in our collection that do that. There are no timeouts associated with portwatcher builds; they are allowed to run indefinitely. Barring a power outage, it should work.

> > At the very least, the watchers are going to be tied up for a LONG time, processing the full dependency list...

Yes, it will take a long time, hours most likely. But I don't think it can be avoided, so if we're doing this change, go ahead and do it.

> So before merging, we might want to break this up into at a few separate commits. (Preferably no more than 100 rev-bumps each.)

That probably won't be very effective, unless you commit a batch of revbumps, wait for all builders to begin working on those, then commit another batch, wait for buildbot to get to that batch, etc. It would also be tedious to try to construct small batches of ports in which there are no dependencies on other ports that use openssl.

Buildbot always updates to the latest code in the macports-ports repository before any build. So if you push 5 commits each at least 5 seconds apart, and assuming buildbot was idle when you did so, then buildbot will get to work on the first batch immediately, but by the time it gets to the second batch, the third through fifth batches are already waiting so the ports tree that the second batch uses is current up to the fifth batch, or whatever subsequent commits came in. If any of the ports in commit 2 depend on any of the ports modified in commits 3-5, those ports will get scheduled in response to commit 2.

In addition, all commits made within 5 seconds of one another are merged into one portwatcher build. This applies for example if you have multiple commits in a single pull request or if you push multiple commits to master manually at once.

I would probably not bother with separate small batches and just do it all at once and accept that subsequent builds will have to wait.

**Comment by RJVB** [comment]
To save others a bit of clicking:

> OpenSSL 3.0 is a major release and consequently any application that currently uses an older version of OpenSSL will at the very least need to be recompiled in order to work with the new version. It is the intention that the large majority of applications will work unchanged with OpenSSL 3.0 if those applications previously worked with OpenSSL 1.1.1. However this is not guaranteed and some changes may be required in some cases. Changes may also be required if applications need to take advantage of some of the new features available in OpenSSL 3.0 such as the availability of the FIPS module. 


> https://wiki.openssl.org/index.php/OpenSSL_3.0#Main_Changes_in_OpenSSL_3.0_from_OpenSSL_1.1.1
> […](#)
> 05.11.2021 09:55 René Bertin ***@***.***>: What I'm missing is a little bit of background.

**Comment by cjones051073** [comment]
> I suppose one could get the rev-bump approval here, but drop the actual change like it's already been done elsewhere AFAIK. If shouldn't take a lot of time to adapt your script so it applies only to approved ports, and applies the changes to your working copy of the actual ports tree. It could even do the pushing at some kind of interval (but I doubt the build bots will be less stressed by a series of commits than a single big commit). What I'm missing is a little bit of background. I wasn't even aware of OpenSSL3 (what ever happened to OpenSSL2?), only vaguely about the OpenSSL PG. I think the least would be to touch a few words on why the rev-bump is necessary. Is it because of an ABI incompatibility or some potentially less stringent reason?

OpenSSL changed their version strategy with version 3. It always kinda followed a MAJOR.MINOR.PATCH scheme, but they only guaranteed ABI compatibility on MAJOR.MINOR. This is why the previous two 'major' versions were 1.0.x and 1.1.x. The other side affect of this was having only the PATCH level left to play with was not enough granularity, which is why we got the weird alpha-numeric patch versions, like 1.1.1k etc.

Now with version 3, ABI compatibility is only on the MAJOR version. So the MINOR can (and will) change whilst remaining ABI compatible. We also (hopefully) will not need the alpha-numeric patch levels any more.

So, to answer your question, there was no version 2 as effectively 1.1.x was the second 'major' version.

**Comment by kimuraw** [comment]
@cjones051073 , all of versions of ruby (ruby/openssl) do not support openssl-3.
please remove lang/ruby* from this PR.

I'll set openssl.branch to 1.1 for ruby24 .. ruby30.

- ruby/openssl do not support version 2.2.x or earlier
  - https://github.com/ruby/openssl/issues/369
- latest version of ruby/openssl is 2.2.1
  - ruby-3.0.2 includes 2.2.0, build fails with port:openssl3

**Comment by cjones051073** [comment]
> @cjones051073 , all of versions of ruby (ruby/openssl) do not support openssl-3. please remove lang/ruby* from this PR.
> 
> I'll set openssl.branch to 1.1 for ruby24 .. ruby30.
> 
> * ruby/openssl do not support version 2.2.x or earlier
>   
>   * [Support OpenSSL 3.0 ruby/openssl#369](https://github.com/ruby/openssl/issues/369)
> * latest version of ruby/openssl is 2.2.1
>   
>   * ruby-3.0.2 includes 2.2.0, build fails with port:openssl3

I cannot 'remove' them from this PR in the sense that this PR updates the default openssl version to 3.

If/when we do make the switch to use openssl3 by default, the ruby ports which do not  support this version will need to be adapted to directly use the openssl PG to peg themselves back to openssl 1.1.

**Comment by cjones051073** [comment]
@kimuraw I have removed the rev-bump of all ruby ports here, so this PR no long touches them. you can update them to directly use the openssl PG to peg themselves to 1.1 as and when you want, it doesn't have to wait for this change.

**Comment by RJVB** [comment]
port:VLC2 should be removed too. It doesn't build currently so there's no point in revbumping.

File: multimedia/mythtv.28/Portfile
Comment (jmroot) [inline multimedia/mythtv.28/Portfile:642]: The mythtv.28 port doesn't need to be bumped but the mythtv-core.28 and mythtv-plugins.28 subports do.

File: net/openssh/Portfile
Comment (jmroot) [inline net/openssh/Portfile:257]: The ssh-copy-id subport doesn't need to be bumped.

File: security/cyrus-sasl2/Portfile
Comment (jmroot) [inline security/cyrus-sasl2/Portfile:9]: It would be nice to bump the revision to 5 and remove the platform darwin 20 block below that increments the revision.

File: security/yara/Portfile
Comment (jmroot) [inline security/yara/Portfile:77]: The py*-yara subports don't need to be bumped.

File: sysutils/john/Portfile
Comment (jmroot) [inline sysutils/john/Portfile:8]: The john port does not need to be bumped (just the john-jumbo subport).

File: www/nghttp2/Portfile
Comment (jmroot) [inline www/nghttp2/Portfile:8]: The nghttp2 port doesn't need to be bumped, but the nghttp2-tools subport does.

**Review by jmroot — CHANGES_REQUESTED** [review]
These Portfiles appear to need a rev bump but are not currently included:
```
databases/mariadb-10.2/Portfile
databases/mariadb-10.3/Portfile
databases/mariadb-10.4/Portfile
databases/mariadb-10.5/Portfile
databases/mariadb-10.6/Portfile
databases/mysql5/Portfile
databases/mysql56/Portfile
databases/percona/Portfile
perl/p5-crypt-openssl-ec/Portfile
perl/p5-crypt-openssl-x509/Portfile
perl/p5-git-raw/Portfile
perl/p5-net-dns-sec/Portfile
perl/p5-net-ssh2/Portfile
```
@dbevans @michaelld @ryandesign

**Comment by kimuraw** [comment]
done at 36878b9a51aa0e184f9c1282c631fbedeac3b05b

ports ruby24 .. ruby30 uses openssl11.

> @cjones051073 , all of versions of ruby (ruby/openssl) do not support openssl-3. please remove lang/ruby* from this PR.
> 
> I'll set openssl.branch to 1.1 for ruby24 .. ruby30.
> 
> * ruby/openssl do not support version 2.2.x or earlier
>   
>   * [Support OpenSSL 3.0 ruby/openssl#369](https://github.com/ruby/openssl/issues/369)
> * latest version of ruby/openssl is 2.2.1
>   
>   * ruby-3.0.2 includes 2.2.0, build fails with port:openssl3

**Comment by gctwnl** [comment]
I do maintain the nsd port (updates etc.) but this is the first time I have seen such a request as this.

I have a local clone of my own fork, so my guess s something like

- create a git branch for this in my own clone of my own fork
- git pull upstream name-of-branch-that-has-this-PR
- compile nsd (I don’t know of nsd has a testing suite that can run, nor do I use SSL-like stuff in my own nsd setup)

Is that it?

Gerben Wierda (LinkedIn <https://www.linkedin.com/in/gerbenwierda>)
R&A Enterprise Architecture <https://ea.rna.nl/> (main site)
Book: Chess and the Art of Enterprise Architecture <https://ea.rna.nl/the-book/>
Book: Mastering ArchiMate <https://ea.rna.nl/the-book-edition-iii/>

**Comment by cjones051073** [comment]
> I do maintain the nsd port (updates etc.) but this is the first time I have seen such a request as this. I have a local clone of my own fork, so my guess s something like - create a git branch for this in my own clone of my own fork - git pull upstream name-of-branch-that-has-this-PR - compile nsd (I don’t know of nsd has a testing suite that can run, nor do I use SSL-like stuff in my own nsd setup) Is that it? Probably not.

Depends what you aim is ? What do you have in min d doing ?

**Comment by cjones051073** [comment]
> These Portfiles appear to need a rev bump but are not currently included:
> 
> ```
> databases/mariadb-10.2/Portfile
> databases/mariadb-10.3/Portfile
> databases/mariadb-10.4/Portfile
> databases/mariadb-10.5/Portfile
> databases/mariadb-10.6/Portfile
> databases/mysql5/Portfile
> databases/mysql56/Portfile
> databases/percona/Portfile
> perl/p5-crypt-openssl-ec/Portfile
> perl/p5-crypt-openssl-x509/Portfile
> perl/p5-git-raw/Portfile
> perl/p5-net-dns-sec/Portfile
> perl/p5-net-ssh2/Portfile
> ```
> 
> @dbevans @michaelld @ryandesign

Thanks! These look like portfiles that set the revision in 'non standard' ways and the script I used was unable to handle them.

**Comment by cjones051073** [comment]
> done at [36878b9](https://github.com/macports/macports-ports/commit/36878b9a51aa0e184f9c1282c631fbedeac3b05b)
> 
> ports ruby24 .. ruby30 uses openssl11.
> 
> > @cjones051073 , all of versions of ruby (ruby/openssl) do not support openssl-3. please remove lang/ruby* from this PR.
> > I'll set openssl.branch to 1.1 for ruby24 .. ruby30.
> > 
> > * ruby/openssl do not support version 2.2.x or earlier
> >   
> >   * [Support OpenSSL 3.0 ruby/openssl#369](https://github.com/ruby/openssl/issues/369)
> > * latest version of ruby/openssl is 2.2.1
> >   
> >   * ruby-3.0.2 includes 2.2.0, build fails with port:openssl3

Apologies but I did not notice you had pushed commits here, and just force pushed some changes that wiped them out.

The ruby changes do not need to be tied to this PR, so can I suggest you make a separate PR for those, or just push them directly to master.

**Comment by cjones051073** [comment]
> > done at [36878b9](https://github.com/macports/macports-ports/commit/36878b9a51aa0e184f9c1282c631fbedeac3b05b)
> > ports ruby24 .. ruby30 uses openssl11.
> > > @cjones051073 , all of versions of ruby (ruby/openssl) do not support openssl-3. please remove lang/ruby* from this PR.
> > > I'll set openssl.branch to 1.1 for ruby24 .. ruby30.
> > > 
> > > * ruby/openssl do not support version 2.2.x or earlier
> > >   
> > >   * [Support OpenSSL 3.0 ruby/openssl#369](https://github.com/ruby/openssl/issues/369)
> > > * latest version of ruby/openssl is 2.2.1
> > >   
> > >   * ruby-3.0.2 includes 2.2.0, build fails with port:openssl3
> 
> @kimuraw Apologies but I did not notice you had pushed commits here, and just force pushed some changes that wiped them out.
> 
> The ruby changes do not need to be tied to this PR, so can I suggest you make a separate PR for those, or just push them directly to master.

Ah, my mistake, the changes are in master already....

**Comment by cjones051073** [comment]
@jmroot Thanks for your comments, I hope I have now addressed them all.

**Comment by gctwnl** [comment]
> > I do maintain the nsd port (updates etc.) but this is the first time I have seen such a request as this. I have a local clone of my own fork, so my guess s something like - create a git branch for this in my own clone of my own fork - git pull upstream name-of-branch-that-has-this-PR - compile nsd (I don’t know of nsd has a testing suite that can run, nor do I use SSL-like stuff in my own nsd setup) Is that it? Probably not.
> 
> Depends what your aim is ? What do you have in mind doing ?

Well, I got these messages (in fact I got all these mentions in my email box) so I assumed I had to do something. I'm trying to find out what. If nothing, that's OK too.

(I noticed the nds version could use a bump too, so it was a good reason to look into that and now I'm trying to fix the expired root certificate problem on my macOS Mojave main workhorse, which Apple just dropped from patching and which I cannot migrate until I have time for it late December)

**Comment by RJVB** [comment]
On Saturday November 06 2021 03:43:26 gctwnl wrote:
> I'm trying to fix the expired root certificate problem on my macOS Mojave main workhorse, which Apple just dropped from patching and which I cannot migrate until I have time for it late December)

Check out the updated port:apple-pki-bundle port for that!

**Comment by cjones051073** [comment]
> > > I do maintain the nsd port (updates etc.) but this is the first time I have seen such a request as this. I have a local clone of my own fork, so my guess s something like - create a git branch for this in my own clone of my own fork - git pull upstream name-of-branch-that-has-this-PR - compile nsd (I don’t know of nsd has a testing suite that can run, nor do I use SSL-like stuff in my own nsd setup) Is that it? Probably not.
> > 
> > 
> > Depends what your aim is ? What do you have in mind doing ?
> 
> Well, I got these messages (in fact I got all these mentions in my email box) so I assumed I had to do something. I'm trying to find out what. If nothing, that's OK too.

You get automatic messages when you are mentioned in a PR, e.g. if a port you maintain is touched. It does not necessarily imply you have to do anything specific.

**Comment by mascguy** [comment]
Hey Chris, given that these changes will tie up the CI build jobs until timeout, go ahead and add `[skip ci]` to the commit comment. Documented via the following blog post:

https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/

In the interim, I've been manually cancelling them. But that's a pain. LOL

**Comment by gctwnl** [comment]
> On Saturday November 06 2021 03:43:26 gctwnl wrote: I'm trying to fix the expired root certificate problem on my macOS Mojave main workhorse, which Apple just dropped from patching and which I cannot migrate until I have time for it late December)
> Check out the updated port:apple-pki-bundle port for that!

I wonder how that could help. If the problem is that www.nlnetlabs.nl (the source for nsd) uses a letsencrypt certificate that has a complete chain including the cross-siding of the correct root certificate by the expired root certificate, then changing anything at my end doesn't work.

Instead, I've asked NLnet Labs to update their certificate according the the MacPorts instructions at https://trac.macports.org/wiki/ProblemHotlist#letsencrypt and they will do so.

**Comment by chrstphrchvz** [comment]
Pinging those who might not have been earlier in case of [limit of 50 mentions per comment](https://trac.macports.org/ticket/63821) on behalf of @macportsbot:

> @alfredh for port baresip, libre, librem, libsrtp.
> @michaelld for port qt4-mac, dvisvgm.
> @ryandesign for port dar, minizip-ng, wimlib, faust-devel, faust, faustlive-devel, eet, libhsplasma, libmongo-client, PlasmaClient, bahamut, php, curl, docsis, kerberos5, net-snmp, rabbitmq-c, rsync, shellinabox, php-event, php-mongo, php-stomp, php-swoole, aircrack-ng, openvas-client, rsyncrypto, NetSurf, QupZilla, apache2, lighttpd, slowhttptest, winetricks.
> @jolan78 for port cryfs.
> @ryanakca for port barrier, opensmtpd.
> @markemer for port Io.
> @dgilman for port redis.
> @g5pw for port pev, radare2, rust, esniper, lnav, socat.
> @arjanvandervelde for port poco.
> @davidnich for port qore-asn1-module, qore-json-module, qore-xml-module, qore.
> @danielluke for port bind9, nmap, sysmon, clamav, ntp.
> @lpsinger for port py-twilio, ds9, htcondor, lscsoft-deps.
> @agraef for port faust-devel, faust, faustlive-devel.
> @roederja for port msodbcsql, msodbcsql17, mico, sequoia-pgp.
> @gpanders for port netdata.
> @nareshov for port git-crypt.
> @schiebel for port mico.
> @tobypeterson for port ldmud, epic4, epic5.
> @jandemter for port py-trio, py-trustme.
> @fhgwright for port libaes_siv, ntpsec.
> @jvirkki for port dupd.
> @Schamschula for port xar, xeus-sqlite, xeus, lftp, tcpdump, zeek, py-acme, py-ndg_httpsclient, certbot, stunnel, apache2, goaccess, nghttp2, nginx.
> @mascguy for port openldap-devel, openldap.
> @sikmir for port lagrange.
> @jrjsmrtn for port wpa_passphrase.
> @tenomoto for port nco.
> @Raimondi for port lastpass-cli.
> @toy for port stressdrive.
> @mklein-de for port cadaver.
> @raimue for port gmpc, hexchat, sslscan.
> @dbevans for port opusfile, telepathy-idle, libgda5, libgit2, podofo, deluge, libtorrent-rasterbar, p5-crypt-smime.
> @mohd-akram for port crystal, websocketpp, cpprestsdk.
> @jcvernaleo for port alpine.
> @nano103 for port libp11.
> @amake for port yubikey-manager.
> @catap for port retdec.
> @cardi for port weechat.
> @alexeyt820 for port librdkafka.
> @millerdev for port openfortivpn.
> @bwalle for port crackpkcs12.
> @Gminfly for port libssh2, medusa.
> @NicosPavlov for port kdelibs4, smokekde.
> @danchr for port pypy, p5-net-eboks.
> @essandess for port mail-server, apple-pki-bundle, calendar-contacts-server, privoxy.
> @akkornel for port munge.
> @Jakker for port softhsm.
> @jerryyhom for port libevent, alpine.
> @Veence for port gdal, qgis, qgis3.

**Review by danchr — APPROVED** [review]
I do not expect any particular breakage for PyPy, but I will check the port once it's built. Also: Yes, please! This will matter greatly for the Mercurial ports I also maintain.

**Review by jmroot — APPROVED** [review]

**Review by judaew — APPROVED** [review]
LGTM for my ports

**Comment by lbschenkel** [comment]
I'm happy with whatever is the consensus and if the ports build.

**Comment by jmroot** [comment]
> You get automatic messages when you are mentioned in a PR, e.g. if a port you maintain is touched. It does not necessarily imply you have to do anything specific.

However, you can test if your port builds against openssl 3 by adding `PortGroup openssl 1.0` and in most cases simply setting `openssl.branch 3` (some extra portgroup options may be needed to help some ports find the right openssl version). If it fails to build, you can set `openssl.branch 1.1` instead for the time being, and ask for the port to be excluded from this PR.

**Comment by toy** [comment]
I've managed to check stressdrive against openssl-3, but for that I needed to change include and lib paths in `Portfile` (`${prefix}/include` to `${prefix}/include/openssl-3` and `${prefix}/lib` to `${prefix}/lib/openssl-3`). Do I understand right that with this PR the version of openssl that will be symlinked directly to include and lib will change to version 3?

**Comment by cjones051073** [comment]
> I've managed to check stressdrive against openssl-3, but for that I needed to change include and lib paths in `Portfile` (`${prefix}/include` to `${prefix}/include/openssl-3` and `${prefix}/lib` to `${prefix}/lib/openssl-3`). Do I understand right that with this PR the version of openssl that will be symlinked directly to include and lib will change to version 3?

Please do not hardcode the paths above, but use the proc methods provided in the PG to access the correct include and lib paths. i.e. `[openssl::include_dir]` and `[openssl::lib_dir]`.

But yes, there are two main ways you can use openssl. 

 1. The most common is to use the `openssl` shim port, that installs sym-links to the default openssl version (currently 1.1, to become 3 with this PR)  into the primary macports prefix. 
 2. Use the openssl PG to configure your port to use the isolated build areas, under `libexec`. If you do this, but do not give a version you will use the default, which is the same as in 1. If you do give a version that will be used regardless of whatever the default is. 

Right now without this PR to test against openss3 you have to follow 2., as the default is 1.1. Once this PR is committed though you are correct the default will become 3, so if your port is OK with this you can if you prefer carry on with approach 1. and just use the shim port, with a rev-bump to trigger a rebuild.

**Comment by cjones051073** [comment]
> I'm happy with whatever is the consensus and if the ports build.

I do not guarantee all the ports build as a) many don't build anyway, regardless of the openssl version and b) If a port needs to be pinned back to openssl 1.1 this will need to be done separately to this PR, I am not going to do that here. This can be tested and do before, or after, this PR is merged.

**Comment by cjones051073** [comment]
> Hey Chris, given that these changes will tie up the CI build jobs until timeout, go ahead and add `[skip ci]` to the commit comment. Documented via the following blog post:
> 
> https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/
> 
> In the interim, I've been manually cancelling them. But that's a pain. LOL

I was not aware of that - Useful to know thanks...

**Comment by Schamschula** [comment]
I got clean rebuilds on all my packages.

However, I found two issues unrelated to this PR:

1. zeus-sqlite fails to build under Mojave (and likely any older macOS version) due to an issue with Apple clang and c++-17 (missing the `filesystem` include file)
2. py-ndg_httpsclient is missing the py-setuptools build dependency. I'm not sure how this could have been missed. This has to be a separate commit.

**Comment by cjones051073** [comment]
I think I am going to go ahead and merge this, as given the number of ports it touches I am constantly having to rebase it to fix conflicts as those ports get independent updates. We seem to have approval here and I have not heard anything on the mailing lists objecting. I know it hasn't been a very long period, but we are going to have to hit go on this sometime to switch to the latest openssl so it might as well be now...

**Comment by catap** [comment]
I'm testing both `retdec` ports, but it requires couple of hours to build each. It should work, but let me be sure.

**Comment by RJVB** [comment]
On Sunday November 07 2021 04:38:34 Chris Jones wrote:
>Merged #12807 into master.

No offense, but you *could* have left a little bit more reacting time for those of us who've been too busy with more (sic) urgent matters, esp. on a sunday...

I was under the impression that I was co-maintainer on port:qca (and then only port:qca-qt5) so didn't say anything about it. I'll have to take a look how this is going to work out for the qca*-ossl and -pkcs11 subports.

**Comment by mascguy** [comment]
Geez Chris, thanks for tying up the Buildbot Watchers for hours...  ;-)

Just kidding...

**Comment by ryandesign** [comment]
> @ryandesign for port ... php ...

Looks like the relevant php subports did not end up getting revbumped in this PR after all; I'll do them. afflib was missed too; see f08053a254f3222198b15bf473dbb4b066f3fcd4. I wonder if any others were missed.

**Comment by mascguy** [comment]
Thus far, the failures have been few and far between, so that's good!

There is one notable failure though, for `cargo-c`. It appears that it's being built prior to `openssl` perhaps...?

```
--->  Building cargo-c
DEBUG: system -W /opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cargo-c/cargo-c/work/cargo-c-0.7.3: /opt/local/bin/cargo update
dyld: Library not loaded: /opt/local/libexec/openssl11/lib/libssl.1.1.dylib
  Referenced from: /opt/local/bin/cargo
  Reason: image not found
Command failed: /opt/local/bin/cargo update
```

https://ports.macports.org/port/cargo-c/builds/

The port has a dependency on `openssl`, so I'd expect that to be built first... preventing this issue. @cjones051073 any thoughts/ideas?

CC: @MarcusCalhoun-Lopez

**Comment by mascguy** [comment]
Two more Cargo-dependent ports, exhibiting the same issue (due to a broken openssl dylib on the buildbots):

* amp
* angle-grinder

CC: @herbygillot

**Comment by pmetzger** [comment]
As I mentioned on the bug tracker, ruby30 seems to be broken. It's mysterious because it's part of portgroup openssl 1.0

https://trac.macports.org/ticket/63845

**Comment by onodera-punpun** [comment]
Cargo-c indeed seems to be failing, coincidentally the port is very outdated.

**Comment by cjones051073** [comment]
> Thus far, the failures have been few and far between, so that's good!
> 
> There is one notable failure though, for `cargo-c`. It appears that it's being built prior to `openssl` perhaps...?
> 
> ```
> --->  Building cargo-c
> DEBUG: system -W /opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cargo-c/cargo-c/work/cargo-c-0.7.3: /opt/local/bin/cargo update
> dyld: Library not loaded: /opt/local/libexec/openssl11/lib/libssl.1.1.dylib
>   Referenced from: /opt/local/bin/cargo
>   Reason: image not found
> Command failed: /opt/local/bin/cargo update
> ```
> 
> https://ports.macports.org/port/cargo-c/builds/
> 
> The port has a dependency on `openssl`, so I'd expect that to be built first... preventing this issue. @cjones051073 any thoughts/ideas?
> 
> CC: @MarcusCalhoun-Lopez
> 
> p.s. I'm also seeing a build failure locally, as is at least one of our users. Note that this is definitely not specific to Monterey, as my failure is occuring on 10.13:
> 
> [63848 - cargo-c @0.7.3_1 error: pasting formed 'RUST_VERSION_OPENSSL_](https://trac.macports.org/ticket/63848)

the cargo port needed a rev-bump, it was missed here. I have already submitted that.

**Comment by essandess** [comment]
OpenSSL3 breaks the port `tor` build: https://trac.macports.org/ticket/63870.

**Comment by RJVB** [comment]
So, mostly out of curiosity, now that I'm merging these changes into my own ports tree:

Am I right that there was no immediate need to rev-bump the dependents because they would have continued to find the OpenSSL 1.1.1 libraries in the same locations?

I also have my questions as to the use of an all-in subprefix in libexec with a bunch of symlinks, instead of using `configure` options  to specify the same versioned directories (like lib/openssl-XY) to install just the categories that contain files which have the same name for all opensslXY ports. I've been doing that in my own implementation and it works just fine, but well, whatever :)

**Comment by gctwnl** [comment]
Now that NLNetLabs has fixed their cert I wanted to go on with an NSD update and that also means OpenSSL3 (as that has been released in MacPorts). I started a compile of the existing NSD version.

OpenSSL3 breaks the _existing_ NSD port (v4.2.1) as out has removed ERR_GET_FUNC:

NSD 4.3.8 builds fine. I am testing (lightly) and will prepare a pull request for NSD asap.

**Pull request merged**
